pageserver/src/tenant_tasks.rs

```diff
@@ -0,0 +1,253 @@
+                            for (_, cancel) in gc_loops.drain() {
+                                cancel.send(()).ok();
+                            }
+                            for (_, cancel) in compaction_loops.drain() {
+                                cancel.send(()).ok();
+                            }
+                            // TODO wait for tasks to die?
+                            break;
+                        },
+                        tenantid = gc_recv.recv() => {
+                            let tenantid = tenantid.expect("Gc task channel closed unexpectedly");
+
+                            // Spawn new task, request cancellation of the old one if exists
+                            let (cancel_send, cancel_recv) = watch::channel(());
+                            let handle = tokio::spawn(gc_loop(tenantid, cancel_recv)
+                                .instrument(trace_span!("gc loop", tenant = %tenantid)));
+
+                            // Update metrics, remember handle + cancellation channel
+                            TENANT_TASK_EVENTS.with_label_values(&["start"]).inc();
+                            futures.push(handle);
+                            if let Some(old_cancel_send) = gc_loops.insert(tenantid, cancel_send) {
+                                old_cancel_send.send(()).ok();
+                            }
+                        },
+                        tenantid = compaction_recv.recv() => {
+                            let tenantid = tenantid.expect("Compaction task channel closed unexpectedly");
+
+                            // Spawn new task, request cancellation of the old one if exists
+                            let (cancel_send, cancel_recv) = watch::channel(());
+                            let handle = tokio::spawn(compaction_loop(tenantid, cancel_recv)
+                                .instrument(trace_span!("compaction loop", tenant = %tenantid)));
+
+                            // Update metrics, remember handle + cancellation channel
+                            TENANT_TASK_EVENTS.with_label_values(&["start"]).inc();
+                            futures.push(handle);
+                            if let Some(old_cancel_send) = compaction_loops.insert(tenantid, cancel_send) {
+                                old_cancel_send.send(()).ok();
+                            }
+                        },
+                        result = futures.next() => {
+                            // Log any errors and panics
+                            match result {
+                                Some(Ok(Ok(()))) => {
+                                    TENANT_TASK_EVENTS.with_label_values(&["stop"]).inc();
+                                },
+                                Some(Ok(Err(e))) => {
+                                    TENANT_TASK_EVENTS.with_label_values(&["fail"]).inc();
+                                    // TODO which tenant failed?
```

Comment:
would be great to know. not sure if it's possible with this approach of error reporting